nav2_util/include/nav2_util/raytrace_line_2d.hpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2008, 2013, Willow Garage, Inc.
+// All rights reserved.
+//
+// Software License Agreement (BSD License 2.0)
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above
+//    copyright notice, this list of conditions and the following
+//    disclaimer in the documentation and/or other materials provided
+//    with the distribution.
+//  * Neither the name of Willow Garage, Inc. nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Author: Eitan Marder-Eppstein
+//         David V. Lu!!
+
+#ifndef NAV2_UTIL__RAYTRACE_LINE_2D_HPP_
+#define NAV2_UTIL__RAYTRACE_LINE_2D_HPP_
+
+#include <limits.h>
+#include <algorithm>
+#include <cmath>
+
+namespace nav2_util
+{
```

Comment:
Ditto with all of this; I expect the future data structure we use will have a better implemented raytracing method that I would rather not keep hanging around after we refactor and remove costmap2D

Reply:
I am not sure, that understand it well. Could you please give more context about replacing of Costamp2D? Is is planned to remove&replace all algorithms related to costmaps (including this VO-server/VO-node contribution)?

Reply:
This is still very TBD so I don't have alot of details to share, but Costmap2D is a real rat's nest that needs to be reworked from the ground up with modern data structures and formatting to be more understandable and controllable. Also, it would be great to simultaneously think about AI methods and other data other than `cost` that users might want to store. So, a rework is planned as my principal project of 2024, but concrete details aren't available yet since its not 2024 and I haven't started thinking about the finite details in ... well detail :laughing: 

The VO and filters work will be ported over to work with the new system, those are useful / important capabilities. There will be no loss in capability in the migrartion; only expansion. But it will start with a fresh package rather than building on costmap2D